### PR TITLE
LSIF: Cross-repository database to target Postgres

### PR DESCRIPTION
**Pull request by efritz**
The current LSIF backend uses SQLite as its data store for everything: each repo/commit pair has its own SQLite database (written once by one process and read many times by many processes), as well as a cross-repository database (written by the worker and read many times by many processes).

We will soon have a need for many writers (both the api and worker) to access the cross-repository database concurrently, which will lead to (1) contention on a disks-based database, and (2) higher chance of corruption (journaling modes and exclusive locks still do not solve this problem). This work enables https://github.com/sourcegraph/sourcegraph/pull/5691 to be implemented without chance of data corruption due to multiple writers.

To get ahead of this, this PR splits the data that can be written concurrently by multiple processes to point at Postgres instead of SQLite. In dev, this tablespace can be shared with the same postgres instance used by the frontend/repo-updater/mangement-console. In production, it should likely have its own instance. Tagging distribution (@slimsag) and core service (@tsenart) members that commented on this distribution issue in [RFC 25](https://docs.google.com/document/d/1FVP76WEDt7Y5g9NPz05lz0AO67sP5K9ItX_4eXNcWEg/edit).

### CODE REVIEW

**Comment by tsenart** [comment]
I won't be able to review this PR this week. @keegancsmith or @mrnugget: Would you have the chance?

File: lsif/src/connection.ts [hunk 1]
```diff
@@ -1,24 +1,68 @@
...
     // eslint-disable-next-line @typescript-eslint/ban-types
     entities: Function[]
 ): Promise<Connection> {
+    return _createConnection({ type: 'sqlite', name: database, database, entities, ...baseOptions })
+}
+
+/**
+ * Create a Postgres connection. This creates a typorm connection
+ * pool with the name `xrepo`. The connection configuration is
+ * constructed by `createPostgresConnectionOptions.
```
Comment (mrnugget) [inline lsif/src/connection.ts]:
```suggestion
 * constructed by `createPostgresConnectionOptions`.
```

**Comment by felixfbecker** [comment]
@efritz let me know when you want me to do another review once the architecture/functionality is approved

**Comment by efritz** [comment]
> Regarding Postgres: the idea is to use the same Postgres instance, but a different database, correct? If so, do we have enough documentation in place so that users know to configure two different PGDATASOURCE strings?

Well, that's what I'd like for @sourcegraph/distribution to weigh in on. Conversation in RFCs made me believe that it would be better/safer for a second instance of Postgres, but if another database within the same container works, then that's possible as well.

I'm also looking for advice on all the places where `PG_` variables are set in different contexts so that I can take care of everything in one effort.

**Comment by slimsag** [comment]
Architecture-wise:

1) A separate DB in the same Postgres instance obviously sounds most ideal to me because that means admins don't need to e.g. provision another RDS database. But, it's not clear to me how much load this will place on Postgres so it is hard to know if a separate instance is actually warranted or not. I would suggest rolling this out on a large instance (k8s.sgdev.org) with the separate approach and measuring the change. We could for example take a day to do this and measure how much the Postgres disk grows in size and maybe extract some generic info on avg query time -- which I imagine would probably be enough to gain some confidence (looks like this could maybe be used: https://www.citusdata.com/blog/2019/02/08/the-most-useful-postgres-extension-pg-stat-statements/ )

2) If you do decide going with a separate postgres instance entirely is needed, we would still use just one in the server case I think (no point in running two in the same container).

3) You can just specify the PG_* vars on your lsif-server container. This is:
  - https://github.com/sourcegraph/deploy-sourcegraph/blob/master/base/lsif-server/lsif-server.Deployment.yaml
  - https://github.com/sourcegraph/deploy-sourcegraph-docker/blob/master/deploy-lsif-server.sh
  - https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/server/shared/shared.go#L149-150 (here you _just_ need to override `PGDATABASE` using your own new env var named e.g. `PGDATABASE_LSIF` which users can configure - [the others have preconfigured values](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/server/shared/postgres.go#L89-94))

4) In any case, you should make use of the same tool we use for migrations and create a similar strategy for applying migrations to this DB: see https://github.com/golang-migrate/migrate and https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/migrations/README.md

5) Please please please just make sure the above is documented well :) Every hour you spend making the above seamless, documenting expected increase in resource usage on postgres, etc. saves our admin users several hours -- it's time well spent!

**Comment by tsenart** [comment]
@slimsag's advice is sound architecturally. At the code level, I'm not the most qualified to review TypeScript, so I'll defer.

**Comment by efritz** [comment]
Updated with migrations and docs. I'm looking to see how I can auto-create the lssif database in the base postgres image.

https://github.com/sourcegraph/deploy-sourcegraph/pull/386
https://github.com/sourcegraph/deploy-sourcegraph-docker/pull/44

**Comment by efritz** [comment]
[ ] Remaining tasks: ensure that postgres 9 -> 11 upgrade path can support multiple databases.

File: migrations/1528395594_create_lsif_database.up.sql [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    ELSE
+        -- Get current user as variable
+        SELECT current_user INTO user;
+
+        PERFORM dblink_exec(
+            -- Connect to same database as same user
+            'dbname=' || current_database() || ' user=' || user,
+            -- Create new database as current user
+            'CREATE DATABASE ' || quote_ident(dbname) || ' OWNER ' || user || ' TEMPLATE template0 ENCODING ''UTF8'''
+        );
```
Comment (chrismwendt) [inline migrations/1528395594_create_lsif_database.up.sql]: Does `dblink_exec` somehow tie the query to the enclosing transaction, or is this a backdoor that circumvents the transaction? Do the frontend migrations have to solve this problem, too?

Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: The entire point is to execute it outside of any transaction (we need the function to determine if we should run it or not, which implicitly creates a transaction).

This **is** a frontend migration. The only reason it's necessary at all is because it's creating a new database.

File: lsif/src/connection.ts [hunk 3]
```diff
@@ -1,24 +1,256 @@
...
+ */
+async function getMigrationVersion(connection: Connection): Promise<string> {
+    const rows = (await connection.query(`select * from ${schemaMigrationTableName}`)) as {
+        version: string
+        dirty: string
+    }[]
+
+    if (rows.length > 0) {
+        if (rows[0].dirty) {
+            throw new Error('Current migration is dirty.')
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/connection.ts]: I'll put a comment that this should actually not ever happen - it can't be dirty if we're this far in the app as the migrate tool would first blow up. This mirrors the logic on the Go side (that has the same impossibility note).

File: lsif/src/connection.ts [hunk 4]
```diff
@@ -1,24 +1,256 @@
...
+        return false
+    }
+
+    for (const message of ['no migration found for version', 'file does not exist']) {
+        if (!error.message.includes(message)) {
+            return false
+        }
+    }
+
+    return true
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/connection.ts]: They're not mutually exclusive, they should *both* be there. Reworked to clarify.

File: lsif/src/connection.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/connection.ts]: Since it's unreachable I'm just removing the check

File: lsif/src/connection.ts [hunk 5]
```diff
@@ -1,24 +1,256 @@
...
+        url.searchParams.set('sslmode', 'disable')
+    }
+
+    return url.href
+}
+
+/**
+ * Call the `migrate` tool and log how many migrations were applied. This
+ * blocks on an advisory lock, so running this tool in multiple contexts
+ * should not cause concurrency problems.
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/connection.ts]: Simplified wording.

**Comment by efritz** [comment]
> When is concurrent write access needed in both the server and worker, and why don't journaling modes and exclusive locks solve this problem?

Because we need to update commit data on demand, the API needs also write access to the xrepo db (alternatively, we could drop that request and queue the update, which is bad UX, or we could queue a job and block until it's complete, which would cause long delays if there are any LSIF uploads to process). 

SQLite requires exclusive write access (an exclusive write file lock) on the database for any write. For one, this could be slow with multiple API and/or workers trying to update things. But more importantly, SQLite is basically a wrapper around filesystem calls. Locks in SQLite are implemented in terms of POSIX file locking. This seems very brittle when the drive is networked and/or shared. I've seen some cautionary tales about data loss when SQLite was written to concurrently via the network, and loosing this database will nerf all of the LSIF data that had been already uploaded/processed.

That's not a failure mode I want to bring to ops, and that's not a customer issue I'd want to bring to any of us (sorry all your code intel is gone, rerun things in CI).

File: migrations/1528395594_create_lsif_database.up.sql [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    END IF;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Create database
+SELECT create_db('lsif');
```
Comment (efritz) [inline migrations/1528395594_create_lsif_database.up.sql]: Should this be parameterized somehow by the app? Or is it ok to have this default? What if a user wants to change it (they can configure the sg/sourcegraph dbname, is it weird not to make this one configurable as well)?

Comment (slimsag) [reply migrations/1528395594_create_lsif_database.up.sql]: I would prefix this with `sourcegraph_` just for name-spacing sake, but our docs on external DB usage clearly indicate we own the entire Postgres instance and if you intend to use postgres for other things you should use a separate instance entirely -- so I do not think this needs to be configurable.

Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: That sounds cleaner.

File: migrations/1528395594_create_lsif_database.up.sql [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    -- Determine if database exists, if so nothing to do
+    IF EXISTS (SELECT 1 FROM pg_database WHERE datname = dbname) THEN
+        RAISE NOTICE 'Database already exists';
+    ELSE
+        -- Get current user as variable
+        SELECT current_user INTO user;
+
+        PERFORM dblink_exec(
+            -- Connect to same database as same user
+            'dbname=' || current_database() || ' user=' || user,
```
Comment (slimsag) [inline migrations/1528395594_create_lsif_database.up.sql]: It looks like you can drop `user` entirely here and it will use the current user as the default

**Comment by slimsag** [comment]
Yeah, as a frontend-based migration that lsif-server waits on we handle all of the following cases nicely -- they are all automatic and handled just as any other DB migration we've had in the past effectively:

1) A new cluster deployment running `sourcegraph/postgres-11.1`
2) An existing cluster deployment running `sourcegraph/postgres-11.1` and upgrading to this new Sourcegraph version
3) A new cluster deployment with external postgres DB v9.6+
4) An existing cluster deployment using an external Postgres DB v9.6+ upgrading.
5) A new `sourcegraph/server` deployment with the built-in postgres
6) An existing `sourcegraph/server` deployment with the built-in postgres upgrading to this new Sourcegraph version
7) A new `sourcegraph/server` deployment with external postgres v9.6+
8) An existing `sourcegraph/server` deployment with external postgres v9.6+ upgrading
9) Dev environments

File: migrations/1528395594_create_lsif_database.up.sql (hunk 2 above)
Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: Well, now EVERYTHING is a frontend migration!

File: migrations/1528395594_create_lsif_database.up.sql (hunk 7 above)
Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: It's the system default, not the current user.

File: migrations/1528395594_create_lsif_database.up.sql [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+-- a IF NOT EXISTS variant of the statement. We need to use dblink to
+-- issue the CREATE DATABASE statement after we confirm that the target
+-- database doesn't already exist. We cannot perform the statement
+-- directly within a transaction, function, or DO block.
+CREATE OR REPLACE FUNCTION create_db(dbname text) RETURNS void AS $$
+DECLARE user varchar;
+BEGIN
+    -- Determine if database exists, if so nothing to do
+    IF EXISTS (SELECT 1 FROM pg_database WHERE datname = dbname) THEN
+        RAISE NOTICE 'Database already exists';
```
Comment (slimsag) [inline migrations/1528395594_create_lsif_database.up.sql]: @efritz we can still get rid of the "if exists" here and ad DROP DATABASE (done via `dblink_exec`) in the down migration.

Comment (slimsag) [reply migrations/1528395594_create_lsif_database.up.sql]: (if you want to simplify this further)

**Comment by slimsag** [comment]
Let me know when this is ready for re-review.

File: migrations/1528395594_create_lsif_database.up.sql (hunk 8 above)
Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: Done.

File: doc/admin/external_database.md [hunk 9]
```diff
@@ -9,6 +9,8 @@ Sourcegraph by default provides its own PostgreSQL and Redis databases for data
 
 You can use your own PostgreSQL server with Sourcegraph if you wish. For example, you may prefer this if you already have existing backup infrastructure around your own PostgreSQL server.
 
+Prior to 3.9, Sourcegraph used a single database within PostgreSQL. As of version 3.9, it requires a second database, `sourcegraph_lsif`, which is automatically created as part of the frontend migration. The name of this database is non-configurable. This second database is a sibling to the original in the same PostgreSQL instance. This necessitates that the configured PostgreSQL user has sufficient grants to perform the `CREATE DATABASE` query.
```
Comment (slimsag) [inline doc/admin/external_database.md]:
```suggestion
Prior to 3.9, Sourcegraph used a single database within PostgreSQL. As of version 3.9, it uses a second database, `sourcegraph_lsif`, which is automatically created upon upgrading. Both databases are in the same Postgres instance, which should be dedicated to just Sourcegraph. The configured PostgreSQL user must have sufficient permission to perform the `CREATE DATABASE` query.
```

Comment (slimsag) [reply doc/admin/external_database.md]: Also copy this line into the CHANGELOG.md

File: doc/dev/local_development.md [hunk 10]
```diff
@@ -179,6 +179,10 @@ You need a fresh Postgres database and a database user that has full ownership o
     [envdir]: https://cr.yp.to/daemontools/envdir.html
     [dotenv]: https://github.com/joho/godotenv
 
+    Note: Sourcegraph will create a secondary database in the same PostgreSQL
+    instance called `sourcegraph_lsif`. It is assumed the PostgreSQL instance
+    is dedicated to Sourcegraph solely.
```
Comment (slimsag) [inline doc/dev/local_development.md]: I pushed a commit to change this wording a bit

File: doc/admin/external_database.md (hunk 9 above)
Comment (efritz) [reply doc/admin/external_database.md]: The entire paragraph?

File: doc/dev/local_development.md (hunk 10 above)
Comment (efritz) [reply doc/dev/local_development.md]: Ok.

Comment (slimsag) [reply doc/dev/local_development.md]: The reasons I did this:

1. In local dev, nobody is specifying the PG env vars per-service, so they don't need to worry about that aspect.
2. It being on the same host doesn't matter to the person reading this document, since they wouldn't know how to specify a different one for that service anyway.

File: lsif/README.md [hunk 11]
```diff
@@ -6,6 +6,10 @@ Sourcegraph receives and stores LSIF files uploaded using [upload.sh](upload.sh)
...
 
+## Database Configuration
+
+The LSIF processes store most of its data in SQLite repositories on a shared disk that are written once by a worker on LSIF dump upload, and read many times by the APIs to answer LSIF/LSP queries. Cross-repository and cross-commit data is stored in Postgres, as this database requires many concurrent writer (which is an unsafe operation for SQLite in a networked application). The Sourcegraph server reads PostgreSQL connection configuration from the [`PG*` environment variables](http://www.postgresql.org/docs/current/static/libpq-envars.html). These value should be configured to point to the **frontend** database. The LSIF processes will connect to the database named `sourcegraph_lsif`, regardless of configuration, but require access to the frontend database to check the status of migration on application startup. Although it currently resides on the same Postgres instance (pod, container, or physical machine depending on the deployment context), separation of this data allows the LSIF database to be easily moved to a completely separate Postgres instance if we find that LSIF access and write patterns put too much stress on a single node.
```
Comment (slimsag) [inline lsif/README.md]:
```suggestion
The LSIF processes store most of its data in SQLite repositories on a shared disk that are written once by a worker on LSIF dump upload, and read many times by the APIs to answer LSIF/LSP queries. Cross-repository and cross-commit data is stored in Postgres, as this database requires many concurrent writer (which is an unsafe operation for SQLite in a networked application). The Sourcegraph server reads PostgreSQL connection configuration from the [`PG*` environment variables](http://www.postgresql.org/docs/current/static/libpq-envars.html). These value should be configured to point to the **frontend** database. The LSIF processes will connect to the database named `sourcegraph_lsif`, regardless of configuration, but require access to the primary Sourcegraph database to check the status of migration on application startup. Although it currently resides on the same Postgres instance (pod, container, or physical machine depending on the deployment context), separation of this data allows the LSIF database to be easily moved to a completely separate Postgres instance if we find that LSIF access and write patterns put too much stress on a single node.
```

Comment (slimsag) [reply lsif/README.md]: I changed the wording here that said "frontend database" to instead say "Sourcegraph's primary database" -- it's not frontend-specific it's just the DB we use everywhere except in lsif-server.

File: lsif/src/connection.ts [hunk 12]
```diff
@@ -1,24 +1,187 @@
-import { createConnection as _createConnection, Connection } from 'typeorm'
+import { Connection, createConnection as _createConnection } from 'typeorm'
+import { userInfo } from 'os'
+import { entities } from './models.xrepo'
+import { PostgresConnectionCredentialsOptions } from 'typeorm/driver/postgres/PostgresConnectionCredentialsOptions'
+import { readEnvInt } from './util'
+
+/**
+ * The minimum migration version required by this instance of the LSIF process.
```
Comment (slimsag) [inline lsif/src/connection.ts]:
```suggestion
 * The minimum migration version required by this instance of the LSIF process.
 * This should be updated any time a new lsif-server migration is added to the
 * migrations/ directory, as we watch the DB to ensure we're on at least this
 * version prior to making use of the DB (which the frontend may still be
 * migrating).
```

File: lsif/src/connection.ts [hunk 13]
```diff
@@ -1,24 +1,187 @@
...
+
+/**
+ * How long to wait between queries to check the current database migration version on startup.
+ */
+const SCHEMA_POLL_INTERVAL = readEnvInt('SCHEMA_POLL_INTERVAL', 5)
+
+/**
+ * How many times to try to connect to the cross-repository database on startup.
+ */
+const MAX_CONNECTION_RETRIES = readEnvInt('MAX_CONNECTION_RETRIES', 12)
```
Comment (slimsag) [inline lsif/src/connection.ts]: If I understand correctly, we'll only wait 12s for the DB to be initialized and fully migrated with this?

I think this is too short, we've seen in some cases with external DBs with high latency that it's taken longer than 30s in some cases: https://github.com/sourcegraph/sourcegraph/issues/4097

I suggest ensuring it can wait for ~5m just to be on the safe side.

Comment (efritz) [reply lsif/src/connection.ts]: That's just the number of retries, there's an interval between requests as well.

File: lsif/src/connection.ts [hunk 14]
```diff
@@ -1,24 +1,187 @@
...
+            const currentVersion = await getMigrationVersion(connection, database)
+
+            // Check to see if the current version is at least the minimum version
+            if (parseInt(currentVersion, 10) >= MINIMUM_MIGRATION_VERSION) {
+                return
+            }
+
+            console.log(`waiting for migrations to be applied (${currentVersion} < ${MINIMUM_MIGRATION_VERSION})`)
+        } catch (error) {
+            console.log('failed to determine current database migration state', error)
```
Comment (slimsag) [inline lsif/src/connection.ts]: this could log a few times on new servers, because the schema_migrations table might not exist in such cases. Might make sense to only log this after a few seconds

File: lsif/src/connection.ts [hunk 15]
```diff
@@ -1,24 +1,187 @@
...
+
+        // snooze for a bit then retry
+        await new Promise(resolve => setTimeout(resolve, SCHEMA_POLL_INTERVAL * 1000))
+    }
+}
+
+/**
+ * Gets the current migration version from the frontend database. Throws on query
+ * error, if no migration version can be found, or if the current migration state
+ * is dirty.
```
Comment (slimsag) [inline lsif/src/connection.ts]: document why we use dblink here?

File: lsif/src/connection.ts (hunk 14 above)
Comment (efritz) [reply lsif/src/connection.ts]: That can't be as the connection is already established to the sourcegraph_lsif database. As that's created by a migration the schema_migrations table must also necessarily exist.

File: migrations/1528395594_create_lsif_database.up.sql [hunk 16]
```diff
@@ -0,0 +1,12 @@
+-- dblink allows us to query other databases without opening a second connection.
+-- This is used in this context to run migrations for the LSIF database.
+
+CREATE EXTENSION IF NOT EXISTS dblink;
+
+-- We can't run this particular query within a transaction which, for some reason,
+-- postgres/golang-migrate thinks we're in. See the following issue for additional
+-- details: https://github.com/golang-migrate/migrate/issues/284.
```
Comment (slimsag) [inline migrations/1528395594_create_lsif_database.up.sql]:
```suggestion
-- details: https://github.com/golang-migrate/migrate/issues/284.
-- We use dblink to workaround this issue for now.
```

File: migrations/1528395595_lsif_xrepo_init.up.sql [hunk 17]
```diff
@@ -0,0 +1,34 @@
+-- Performs migration in LSIF database
+
+SELECT dblink_exec('dbname=sourcegraph_lsif user=' || current_user, '
+    -- Now THIS is hacking!
+    CREATE EXTENSION IF NOT EXISTS dblink;
```
Comment (slimsag) [inline migrations/1528395595_lsif_xrepo_init.up.sql]: why is this needed? I would've thought we need this prior to the select statement, not inside it (given we don't use it inside)

File: lsif/src/connection.ts (hunk 13 above)
Comment (efritz) [reply lsif/src/connection.ts]: Updated it to retry for 5 minutes (60 attempts, 5 seconds between attempt).

File: migrations/1528395595_lsif_xrepo_init.up.sql (hunk 17 above)
Comment (efritz) [reply migrations/1528395595_lsif_xrepo_init.up.sql]: We need dblink in the other databases as well so it can check migrations. Updating docs here.

File: lsif/src/connection.ts (hunk 13 above)
Comment (slimsag) [reply lsif/src/connection.ts]: got it -- but the interval is 5s so 12*5s = 60s

I would change this to 60 so we retry for up to 5m in 5s intervals.

File: lsif/src/connection.ts [hunk 18]
```diff
@@ -1,24 +1,191 @@
...
+
+/**
+ * How long to wait between queries to check the current database migration version on startup.
+ */
+const SCHEMA_POLL_INTERVAL = readEnvInt('SCHEMA_POLL_INTERVAL', 5)
+
+/**
+ * How many times to try to connect to the cross-repository database on startup.
+ */
+const MAX_CONNECTION_RETRIES = readEnvInt('MAX_CONNECTION_RETRIES', 12)
```
Comment (slimsag) [inline lsif/src/connection.ts]:
```suggestion
const MAX_CONNECTION_RETRIES = readEnvInt('MAX_CONNECTION_RETRIES', 60)
```

File: lsif/src/connection.ts (hunk 14 above)
Comment (slimsag) [reply lsif/src/connection.ts]: Ahh, got it, I missed that.

File: lsif/src/connection.ts (hunk 18 above)
Comment (efritz) [reply lsif/src/connection.ts]: Beat you to it!

File: migrations/1528395594_create_lsif_database.up.sql (hunk 16 above)
Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: Updated.

File: doc/admin/external_database.md (hunk 9 above)
Comment (slimsag) [reply doc/admin/external_database.md]: I pushed a commit to add an entry just now with what I would write -- feel free to adjust as you see fit.

File: lsif/src/connection.ts (hunk 15 above)
Comment (efritz) [reply lsif/src/connection.ts]: Done.

File: lsif/README.md [hunk 19]
```diff
@@ -6,6 +6,10 @@ Sourcegraph receives and stores LSIF files uploaded using [upload.sh](upload.sh)
...
+
+The LSIF processes store most of its data in SQLite repositories on a shared disk that are written once by a worker on LSIF dump upload, and read many times by the APIs to answer LSIF/LSP queries. Cross-repository and cross-commit data is stored in Postgres, as this database requires many concurrent writer (which is an unsafe operation for SQLite in a networked application). The Sourcegraph server reads PostgreSQL connection configuration from the [`PG*` environment variables](http://www.postgresql.org/docs/current/static/libpq-envars.html). These value should be configured to point to the same Postgres database as the rest of Sourcegraph. The LSIF processes will connect to the database named `sourcegraph_lsif`, regardless of configuration, but require access to the primary Sourcegraph database to check the status of migration on application startup. Although it currently resides on the same Postgres instance (pod, container, or physical machine depending on the deployment context), separation of this data allows the LSIF database to be easily moved to a completely separate Postgres instance if we find that LSIF access and write patterns put too much stress on a single node.
```
Comment (keegancsmith) [inline lsif/README.md]: why does it make it easier? Doesn't the code itself not care what other tables exist in the database? IE if we do decide to move to another PG instance it won't matter if its in the same DB or not currently, it will be the same amount of work to migrate data?

File: lsif/src/connection.ts [hunk 20]
```diff
@@ -1,24 +1,195 @@
...
+
+    return {
+        host: process.env.PGHOST || '127.0.0.1',
+        port: process.env.PGPORT ? parseInt(process.env.PGPORT, 10) : 5432,
+        username: process.env.PGUSER || userInfo().username || 'postgres',
+        password: process.env.PGPASSWORD,
+        database: process.env.PGDATABASE || 'sourcegraph',
+        ssl: process.env.PGSSLMODE === '' ? false : undefined,
+    }
+}
```
Comment (keegancsmith) [inline lsif/src/connection.ts]: FYI other services (like repo-updater) rely on the config returned from Sourcegraph frontend to learn how to connect. This will introduce a new service which depends on configuring these environment variables (not an issue for server, but for clusters). Do we have a path forward for this service to also speak to sourcegraph-frontend's configuration endpoint? (I suppose this is a reason why sticking to go is nice, due to the client code already existing for this). FYI it should be relatively straightforward to implement.

File: CHANGELOG.md [hunk 21]
```diff
@@ -19,6 +19,7 @@ All notable changes to Sourcegraph are documented in this file.
 
 - Allow single trailing hyphen in usernames and org names [#5680](https://github.com/sourcegraph/sourcegraph/pull/5680)
 - Indexed search won't spam the logs on startup if the frontend API is not yet available. [zoekt#30](https://github.com/sourcegraph/zoekt/pull/30)
+- Sourcegraph now creates a secondary database within the configured PostgreSQL instance called `sourcegraph_lsif`. This is done automatically, but please ensure the configured PostgreSQL user has permission to perform `CREATE DATABASE` if using an external database before upgrading.
```
[the remark this answers is not in the document]
Comment (efritz) [reply CHANGELOG.md]: I'm not sure, but I'd like to entertain as much discussion about this as possible (I'm still learning how things are configured for different customers). What kind of issue were you thinking here? We make the assumption that sourcegraph has access to the entire instance. @slimsag added the following in the dev/postgres.md file:

```Note: Sourcegraph will create a secondary database in the same PostgreSQL
instance called `sourcegraph_lsif`. It is assumed the PostgreSQL instance
is dedicated to Sourcegraph solely.```

Maybe we should also note this in the external database docs?

File: lsif/src/connection.ts (hunk 20 above)
Comment (efritz) [reply lsif/src/connection.ts]: Repo-updater reads an HTTP endpoint from the frontend currently to get config? Could you show me where, or am I misunderstanding?

File: lsif/README.md (hunk 19 above)
Comment (efritz) [reply lsif/README.md]: My thought was: (1) bulkhead the data that we _may_ need to move in the future. It may be fine for most or all customers to reside on the same physical instance forever - this needs to be determined via usage patterns; (2) if we need to move it, we can add the necessary code to quickly allow lsif-server/lsif-worker to point to a remote second database (not on the first host). This is a minimal change; (3) pg_dump and pg_restore can operate on the database without having to cherry-pick tables from the original database, which should be a very straightforward ops procedure. I feel like this may be more difficult if we just add the tables to the same db namespace, but I could be wrong here.

File: lsif/src/connection.ts (hunk 20 above)
Comment (slimsag) [reply lsif/src/connection.ts]: It does.. it also handles postgres config changing. This is a nice property to have and should be straightforward to implement, but we can do it in a separate PR.

File: CHANGELOG.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply CHANGELOG.md]: Yes, please add note of this in https://github.com/sourcegraph/deploy-sourcegraph/blob/master/docs/configure.md#configure-custom-postgresql as well

[the remark this answers is not in the document]
Comment (slimsag) [reply CHANGELOG.md]: I don't think this will cause issues for customers with external postgres, though.

File: lsif/src/connection.ts (hunk 20 above)
Comment (efritz) [reply lsif/src/connection.ts]: I will open an issue and punt for now.

File: lsif/README.md [hunk 22]
```diff
@@ -6,6 +6,10 @@ Sourcegraph receives and stores LSIF files uploaded using [upload.sh](upload.sh)
...
...its data in SQLite repositories on a shared disk that are written once by a worker on LSIF dump upload, and read many times by the APIs to answer LSIF/LSP queries. Cross-repository and cross-commit data is stored in Postgres, as this database requires many concurrent writer (which is an unsafe operation for SQLite in a networked application). The LSIF processes read PostgreSQL connection configuration from the [`PG*` environment variables](http://www.postgresql.org/docs/current/static/libpq-envars.html). These value should be configured to point to the same Postgres database as the rest of Sourcegraph. The LSIF processes will connect to the database named `{PGDATABASE}_lsif`, where `{PGDATABASE}` is the primary Sourcegraph database. The processes also require access to the primary Sourcegraph database to check the status of migration on application startup. Although it currently resides on the same Postgres instance (pod, container, or physical machine depending on the deployment context), separation of this data allows the LSIF database to be easily moved to a completely separate Postgres instance if we find that LSIF access and write patterns put too much stress on a single node.
```
Comment (slimsag) [inline lsif/README.md]: You should add a section to this README about where to add a migration and the fact that the const needs to be updated when you do.

File: CHANGELOG.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (efritz) [reply CHANGELOG.md]: https://github.com/sourcegraph/deploy-sourcegraph/pull/386/files#diff-91cd69df8ac1d751bbe7c7b38b6cab1aR488

Anywhere else?

File: migrations/1528395594_create_lsif_database.up.sql [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+
+-- Create the LSIF database. We can't run this particular query within a
+-- transaction which, for some reason, postgres and/or golang-migrate thinks
+-- we're in. See https://github.com/golang-migrate/migrate/issues/284 for
+-- additional details. To issue this query in a non-transactional context,
+-- we use dblink.
+
+CREATE OR REPLACE FUNCTION create_lsif_db() RETURNS void AS $$
+BEGIN
+    -- Buildkite (in particular) will fail if we aren't able to run thiss migration
```
Comment (slimsag) [inline migrations/1528395594_create_lsif_database.up.sql]:
```suggestion
    -- Buildkite (in particular) will fail if we aren't able to run this migration
```

File: lsif/README.md (hunk 22 above)
Comment (efritz) [reply lsif/README.md]: Updated.

File: migrations/1528395594_create_lsif_database.up.sql (hunk 23 above)
Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: https://imagesvc.timeincapp.com/v3/fan/image?url=https%3A%2F%2Fwizardsandwhatnot.com%2Ffiles%2F2017%2F08%2Fsnakes-of-harry-potter.jpg&c=sc&w=736&h=485

File: migrations/1528395594_create_lsif_database.up.sql [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    -- in an idempotent manner. We need to ensure that the database has not already
+    -- been created before issuing the create database command.
```
Comment (slimsag) [inline migrations/1528395594_create_lsif_database.up.sql]: Just to be clear, we still don't know why exactly buildkite is relying on this being idempotent?

If yes, can you file an issue for us to follow up on? This is most likely a big problem

If no, can you rephrase this comment to indicate why there is an idempotent constraint on buildkite here? It's not clear to me

Comment (efritz) [reply migrations/1528395594_create_lsif_database.up.sql]: I'm simplifying the comment and looking into it more.

File: CHANGELOG.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply CHANGELOG.md]: Nope, that's it.

File: lsif/README.md (hunk 19 above)
Comment (keegancsmith) [reply lsif/README.md]: Thanks for the detailed response. I agree, lets move forward with a seperate DB.

> (1) bulkhead the data that we _may_ need to move in the future. It may be fine for most or all customers to reside on the same physical instance forever - this needs to be determined via usage patterns;

Yeah to me this is the best reason to do this. So easy to accidently introduce a dependence otherwise at a later stage.

> (2) if we need to move it, we can add the necessary code to quickly allow lsif-server/lsif-worker to point to a remote second database (not on the first host). This is a minimal change; 

We will also need to adjust migrations. What do you envision here?

> (3) pg_dump and pg_restore can operate on the database without having to cherry-pick tables from the original database, which should be a very straightforward ops procedure. I feel like this may be more difficult if we just add the tables to the same db namespace, but I could be wrong here.

Yeah agreed.

File: lsif/src/connection.ts (hunk 20 above)
Comment (keegancsmith) [reply lsif/src/connection.ts]: Link to issue? @slimsag do any customers use custom PG settings in Sourcegraph cluster? If not, then we can punt on this issue. If so, you either need to document they need to update the lsif k8s resource with the pg settings, or we implement reading from the config.

File: lsif/README.md (hunk 19 above)
Comment (efritz) [reply lsif/README.md]: > We will also need to adjust migrations. What do you envision here?

The frontend migrations targetting the lsif database already use dblink_exec to run the queries in a different database. We can just add a host parameter to make it a remote execution. We'd also need a way to specify a different host in the lsif server, but at that point we would read the config from the frontend which would already have it configured to run the migrations properly.

I feel like it should be a simple (technical) change, but we shouldn't invest effort into it until it's necessary to move the data (which may not ever need to happen).

File: lsif/src/connection.ts (hunk 20 above)
Comment (efritz) [reply lsif/src/connection.ts]: https://github.com/sourcegraph/sourcegraph/issues/5853

Comment (efritz) [reply lsif/src/connection.ts]: And implemented in https://github.com/sourcegraph/sourcegraph/pull/5871
